src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,568 @@
+{
+  return out << "cbj_header_t(magin=" << header.magic
+	     << ", uuid=" << header.uuid
+	     << ", block_size=" << header.block_size
+	     << ", size=" << header.size
+	     << ", journal_tail=" << header.journal_tail
+	     << ", applied_to="<< header.applied_to
+	     << ", written_to=" << header.written_to
+	     << ", flsg=" << header.flag
+	     << ", header_checksum=" << header.header_checksum
+             << ")";
+}
+
+
+CircularBoundedJournal::CircularBoundedJournal(NVMeBlockDevice* device,
+    const std::string path)
+  : device(device), path(path) {}
+
+CircularBoundedJournal::mkfs_ret
+CircularBoundedJournal::mkfs(const mkfs_config_t& config)
+{
+  LOG_PREFIX(CircularBoundedJournal::mkfs);
+  return _open_device(path
+  ).safe_then([this, config, FNAME]() mutable -> mkfs_ret {
+    assert(static_cast<seastore_off_t>(config.block_size) ==
+      device->get_block_size());
+    ceph::bufferlist bl;
+    CircularBoundedJournal::cbj_header_t head;
+    head.block_size = config.block_size;
+    head.size = config.total_size - device->get_block_size();
+    head.journal_tail = device->get_block_size();
+    head.written_to = head.journal_tail;
+    head.applied_to = head.journal_tail;
+    head.device_id = config.device_id;
+    start_dev_addr = config.start;
+    encode(head, bl);
+    header = head;
+    DEBUG(
+      "initialize header block in CircularBoundedJournal, length {}",
+      bl.length());
+    return write_header(
+    ).handle_error(
+      mkfs_ertr::pass_further{},
+      crimson::ct_error::assert_all{
+      "Invalid error device_write during CircularBoundedJournal::mkfs"
+    }).safe_then([]() {
+      return mkfs_ertr::now();
+    });
+  }).safe_then([this]() {
+    if (device) {
+      return device->close(
+      ).safe_then([]() {
+	return mkfs_ertr::now();
+      });
+    }
+    return mkfs_ertr::now();
+  });
+}
+
+CircularBoundedJournal::open_for_write_ertr::future<>
+CircularBoundedJournal::_open_device(const std::string path)
+{
+  ceph_assert(device);
+  return device->open(path, seastar::open_flags::rw
+  ).handle_error(
+    open_for_write_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+      "Invalid error device->open"
+    }
+  );
+}
+
+ceph::bufferlist CircularBoundedJournal::encode_header()
+{
+  bufferlist bl;
+  encode(header, bl);
+  return bl;
+}
+
+CircularBoundedJournal::open_for_write_ret CircularBoundedJournal::open_for_write()
+{
+  return open_for_write(CBJOURNAL_START_ADDRESS);
+}
+
+CircularBoundedJournal::close_ertr::future<> CircularBoundedJournal::close()
+{
+  return write_header(
+  ).safe_then([this]() -> close_ertr::future<> {
+    init = false;
+    return device->close();
+  }).handle_error(
+    open_for_write_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+      "Invalid error write_header"
+    }
+  );
+}
+
+CircularBoundedJournal::open_for_write_ret
+CircularBoundedJournal::open_for_write(rbm_abs_addr start)
```

Comment:
This will not be used interchangeably with open_for_write(), rename it to open_device_read_header.